Add nth percentile calculation

wasnotrice opened the pull request:
Hi all,

Here's a first pass at adding nth percentile calculations for initial review (see #48). Essentially just:

- a percentile function with tests (calculates any percentile value)
- replacing the current median function with the new percentile function (since the median `is` the 50th percentile value)

There are lots of methods for calculating percentile, and I don't know much about the theory. Of the three most accepted, I chose the one that was explained in greatest detail. I left some resources in a comment for further reading, but I feel confident this method is suitable for our purposes, but it would be trivial to replace, or even offer alternate strategies.

This implementation has similar big-O complexity to the current median function (sorting and then traversing the list of samples), although it does remove one list traversal.

Questions:

- Should percentile be included by default, or enabled with an option?

- The issue suggests that we should only offer the 99th percentile. That seems like a reasonable default, although I suspect that folks will have widely varying ideas about what percentile value they are interested in, and it probably depends on their data. Do we want to offer more flexible configuration for percentile? There's nothing to do differently, calculation-wise, only figuring out a good user experience. Could also be added later.

Left to do:

- [ ] add percentile value(s) to results
- [ ] update output tests

devonestes commented on lib/benchee/statistics.ex:
```diff
@@ -159,7 +160,7 @@ defmodule Benchee.Statistics do
     standard_dev_ratio  = deviation / average
     standard_dev_ips    = ips * standard_dev_ratio
-    median              = compute_median(run_times, iterations)
+    median              = Percentile.percentile(run_times, iterations, 50)
```
Good call re-using this! I never would have put that together myself.

wasnotrice replied:
Hahaha, I think you would have. I didn't realize until I kept seeing the same values for median and 50th percentile. And reading!

PragTob replied:
it's non obvious yeah but this is how I imagined it. Thanks :grin:

devonestes commented on lib/benchee/statistics/percentile.ex:
```diff
@@ -0,0 +1,86 @@
+
+  ## Examples
+
+  iex> Benchee.Statistics.Percentile.percentile([5, 3, 4, 5, 1, 3, 1, 3], 8, 100)
```
Thank you for the lovely doctests 🎉

PragTob replied:
:tada: :tada: :tada:

devonestes commented on lib/benchee/statistics/percentile.ex:
```diff
@@ -0,0 +1,86 @@
+  end
+
+  @spec percentile(list(number()), integer(), integer()) :: float()
+  def percentile(samples, number_of_samples, percentile_number) when percentile_number > 100 do
```
First off, do we want `percentile/3` to be public? It seems to me like the caller can in all cases use `percentile/2` and be totally cool.

What would you think about failing in this case and the one on line 40, and giving a helpful error message? I can't imagine that a user would give us a percentage of more than 99 or less than 1 on purpose. I think it would be likely that these would be cases of typos, and that by coercing these error cases we might be giving users results they think are strange and hiding their typos when we should be telling them how to fix their typos and get the results they want.

wasnotrice replied:
@devonestes so....the reason I have `percentile/3` is because `Statistics` has both `samples` and `iterations`. I assumed that was to avoid calculating the length of the list over and over again (though I'm not certain about that). So—you're right, `percentile/2` is really the only _necessary_ interface, unless we want to be able to pass in `number_of_samples` (i.e. `iterations`) for efficiency. Have we.....benchmarked? 😁

wasnotrice replied:
> What would you think about failing in this case and the one on line 40, and giving a helpful error message?

Actually that was my first inclination, but then I backed it out. What do you imagine, throwing something like `raise ArgumentError, "bad percentile value: #{bad_value}. Percentile must be greater than 0 and less than 100"`?

PragTob replied:
For performance see #139 - with what is reasonable to expect to hit us (~1 Million) it takes ~4ms on my machine, we might have multiple scenarios so let's say * 10 that gives 40ms which is ok I guess ;) For 10 Million we'd be at about 300ms total according to this calculation. Non noticable for the given scenario user wise.
However, that's not the real reason I reused `iterations` - it just feels bad to me when I have calculated a value in a top level method and then just recalculate the value in a method directly called from that method. Such waste, could just pass it along.
When I eventually extract a library called `statistex` from benchee I think I'd keep the `/3` version because I'd like to give that possibility to people as they might operate on bigger data and it might make a different for them.

Raising an argument error sounds nice! :+1:

wasnotrice replied:
so, not sure where we stand with `percentile/2` vs `percentile/3`. Speed-wise, not a big deal (thanks @PragTob). Actually, `percentile/2` is not used here at all, I just included it because it felt like the best interface. And I think there is a difference between carrying along the precomputed length within a module, versus allowing it across modules. I mean, if you pass the wrong number there, you get wrong results. So, for that reason, I'm going to remove the `percentile/3` version. It's trivial to add back again in the future.

devonestes commented on lib/benchee/statistics/percentile.ex:
```diff
@@ -0,0 +1,86 @@
+    percentile_value(sorted, rank)
+  end
+
+  defp percentile_value(sorted, rank) when trunc(rank) == 0 do
```
I had never seen `trunc` before - very cool!

PragTob commented on lib/benchee/statistics/percentile.ex:
```diff
@@ -0,0 +1,86 @@
+  Calculates the value at the `percentile_number`-th percentile. Think of this as the
+  value below which `percentile_number` percent of the samples lie. For example,
+  if `Benchee.Statistics.Percentile.percentile(samples, 99)` == 123.45,
+  99% of samples are less than 123.45.
```
less than or equal to imo?

wasnotrice replied:
Definitely open to improvements in the documentation. My understanding is that this value we are calculating is a "less than" number. Here's [what Wikipedia says](https://en.wikipedia.org/wiki/Percentile):

> A percentile (or a centile) is a measure used in statistics indicating the value below which a given percentage of observations in a group of observations fall. For example, the 20th percentile is the value (or score) below which 20% of the observations may be found.
>
> The term percentile and the related term percentile rank are often used in the reporting of scores from norm-referenced tests. For example, if a score is at the 86th percentile, where 86 is the percentile rank, it is equal to the value below which 86% of the observations may be found (carefully contrast with in the 86th percentile, which means the score is at or below the value of which 86% of the observations may be found - every score is in the 100th percentile).

PragTob replied:
imo it's or equal. E.g.

When we have 10 values 1, 2, 3, 4, 5, 6, 7, 8, 9, 10 - the 80th percentile would be 8 and 8 is a member of that 80th percentile group so from my understanding it should be less than or equal

wasnotrice replied:
However....

```elixir
iex(2)> Percentile.percentile([1, 2, 3, 4, 5, 6, 7, 8, 9, 10], 80)
8.8
```

PragTob replied:
but.... but... I guess I have to take a look at the calculation again then and the involved values or the algorithm used to obtain them. I had some suspicions about indexes being off somewhere... for my understanding of 80th percentile (aka 80 percent of the data) this should most definitely return 8 - or where am I going wrong?

PragTob replied:
After cross checking this seems to be the right value (tm) but it still appearsto be weird to me. I mean 80% of my values are up until including 8. In this particular case I would have understood if it's like 8.2 or something... 8.8 just plain seems too weird. Need to up my percentile knowledge apparently

PragTob commented on lib/benchee/statistics/percentile.ex:
```diff
@@ -0,0 +1,86 @@
+
+  def percentile(samples, number_of_samples, percentile_number) do
+    sorted = Enum.sort(samples)
+    rank = (percentile_number / 100) * max(0, number_of_samples + 1)
```
why `number_of_samples + 1` - i.e. why the + 1, shouldn't we be good without it?

wasnotrice replied:
@PragTob I had the same thought. But it's part of the formula. everything comes out wrong if you just use `number_of_samples`. Imagine you have a list of 10 items, and you're looking for the 50th percentile (median):

currently: `rank = (50 / 100) * 11 = 5.5`
without `+ 1`: `rank = (50 / 100) * 10 = 5`

So, as-is, we look for the value halfway between the 5th and 6th elements (yay, median). If we remove the `+1`, then we look for the value of the 5th element (boo, not the median).

PragTob replied:
makes sense... for the odd case of 9, 1/2 * 9 would also be 4.5 but we are looking for 5. Thanks.

However.... dum dum dum - we are 0 indexed. So in my example above we should be looking at number 4 not number 5? But maybe that is redeemed through the `drop`... weird.

I think it's ok as the median tests I wrote back in the day + your tests are passing. One of these things where I trust the tests more than my brain :D

PragTob commented on lib/benchee/statistics/percentile.ex:
```diff
@@ -0,0 +1,86 @@
+  #
+  # For more information on interpolation strategies, see:
+  # - https://stat.ethz.ch/R-manual/R-devel/library/stats/html/quantile.html
+  # - http://www.itl.nist.gov/div898/handbook/prc/section2/prc262.htm
```
:+1: information

PragTob commented on lib/benchee/statistics/percentile.ex:
```diff
@@ -0,0 +1,86 @@
+  # - http://www.itl.nist.gov/div898/handbook/prc/section2/prc262.htm
+  defp interpolation_value(lower_bound, upper_bound, rank) do
+    interpolation_weight = rank - trunc(rank)
+    interpolation_weight * (upper_bound - lower_bound)
```
:heart_eyes:

PragTob commented on lib/benchee/statistics/percentile.ex:
```diff
@@ -0,0 +1,86 @@
+
+  defp percentile_value(sorted, rank) do
+    index = trunc(rank)
+    [lower_bound, upper_bound | _] = Enum.drop(sorted, index - 1)
```
at first I was pretty skeptical, but this seems like a great usage of `drop` - thanks! :+1:

PragTob commented on test/benchee/statistics/percentile_test.exs:
```diff
@@ -0,0 +1,26 @@
+    95.1682
+  ]
+
+  @tag run: true
```
we don't need the tag, do we? :)

wasnotrice replied:
😊

PragTob commented:
Ah, forgot to answer questions.

Re when to display:
I like the idea of always displaying 99th percentile alongside median if the output doesn't get too big. Could totally be done in a separate PR though imo. This is good enough to get in imo (yay small PRs!). Otherwise it'd be part of #50 

Re custom percentiles:
Definitely a cool possible future feature I wanted to have. It's a whole other beast though.
* Do we allow people to override our value? So instead of 99th percentile they can go: `nth_percentile: 95` - might be good and easy enough
* Can the user specify multiple ones aka `nth_percentile: [25, 50, 75, 95]` ? Might be useful, but display gets harder and might play into #50 - totally different beast and can be done later imo :)

wasnotrice commented:
@PragTob @devonestes thanks for the review! I've incorporated most of your feedback in this latest commit. As I worked through it a bit, I ended up adding back `percentile/3` as a private function, and saving off the number of samples internally, so that we can calculate multiple percentiles from the same already-sorted list of samples. This already comes in handy so that we don't do a separate sort for median, and lays the foundation for adding other percentiles in the future if need be.

What do you think about using a map to return multiple percentile values?

```elixir
%{50 => 45.67, 99 => 67.89}
```

I also considered a list of tuples:

```elixir
[{50, 45.67}, {99, 67.89}]
```

but the map seemed nicer for reading out a particular percentile, like the median. Wanted to get your feedback before I got too much deeper into teasing out the test expectations.

There are some weird long line breaks (and even a short variable name) in there to pacify Credo 🔨 

Also, you'll notice I added the percentile header to the console output, but I ran into this map question before I got the data in there

PragTob commented:
@wasnotrice definitely as a map :+1: 

Might do review of the other additions later, wifi is broken here :anger: don't wanna stress my mobile Internet too much :)

PragTob commented on lib/benchee/formatters/console.ex:
```diff
@@ -147,10 +148,11 @@ defmodule Benchee.Formatters.Console do
                        @average_width, "average",
-                       @deviation_width, "deviation", @median_width, "median"])
+                       @deviation_width, "deviation", @median_width, "median",
+                       @percentile_width, "P99"])
```
him `P99` seems weird to me, `99th` or so might be better or wdyt?

wasnotrice replied:
I went back and forth on this: `99th percentile` is too long. I chose `P99` because it's actually a thing people use (according to the internet). But they are probably professional statisticians. I like 99th — we just have to be sure to handle 1st, 2nd, and 3rd too

Other options:

- 99th
- 99%
- 99th %
- 99th %ile

I'll extract it to a function so we can change it easily

PragTob replied:
`99%` looks even better to me :+1:

devonestes replied:
Maybe `99th %`? I'll keep thinking of other possibilities, too, but that's the only one that jumped out to me.

PragTob commented on lib/benchee/statistics.ex:
```diff
@@ -68,7 +69,11 @@ defmodule Benchee.Statistics do
+      typical value you see.
+    * percentiles   - a map of percentile ranks. These are the values below
+      which x% of the run times lie. For example, 99% of run times are shorter
+      than the 99th percentile (P99) rank.
```
shorter or equal to :)

PragTob replied:
and apparently I stand corrected

PragTob replied:
apparently I stand corrected

PragTob commented on lib/benchee/statistics.ex:
```diff
@@ -160,7 +167,8 @@ defmodule Benchee.Statistics do
     standard_dev_ips    = ips * standard_dev_ratio
-    median              = Percentile.percentile(run_times, iterations, 50)
+    percentiles         = Percentile.percentile(run_times, [50, 99])
+    median              = Map.get(percentiles, 50)
```
:+1:

PragTob replied:
I prefer `Map.fetch!` though as I like to blow up early :)

wasnotrice replied:
👍

PragTob commented on lib/benchee/statistics/percentile.ex:
```diff
@@ -1,50 +1,67 @@
+  5.0
+
+  iex> Benchee.Statistics.Percentile.percentile([5, 3, 4, 5, 1, 3, 1, 3], [50, 75, 99])
+  %{50 => 3.0, 75 => 4.75, 99 => 5.0}
```
I feel a bit uneasy about having a different return type for one argument versus list arguments. Not saying needs to change just expressing fears. I think from a user's perspective this here is usually nicer but can come unexpected

wasnotrice replied:
Good point. What do  you think about renaming this function `percentiles/2`, always passing in a list of percentiles, and always returning a map. I kind of like that idea.

PragTob replied:
Sounds good to me :+1:

PragTob commented on lib/benchee/statistics/percentile.ex:
```diff
@@ -1,50 +1,67 @@
-  @spec percentile(list(number()), integer()) :: float()
-  def percentile(samples, percentile_number) do
-    percentile(samples, length(samples), percentile_number)
+  @spec percentile(list(number()), number() | list(number())) ::
```
do we need the parentheses? Looks a bit... distracting.

devonestes replied:
We're getting them with the formatter whether we like it or not, so I'd say we keep them. We could use the `[number()]` syntax for the lists to remove a couple parentheses.

wasnotrice replied:
I only added those because I noticed in my elixir code formatter pull requests, the code formatter was doing that. Or I thought so. @devonestes you wrote an article—does the code formatter add parentheses to types? :)

wasnotrice replied:
@devonestes thanks for answering my question......before I asked it!

PragTob replied:
I'm already starting to _"like"_ the formatter... ;P

(seriously though, why do types in type specs have to look like function calls... makes little sense to me, maybe Jose has a rationale somewhere (I'm sure he has, question is just if he wrote it down somewhere))

PragTob commented on lib/benchee/statistics/percentile.ex:
```diff
@@ -1,50 +1,67 @@
+    |> Enum.reduce(%{}, fn percentile_rank, acc ->
+      perc = percentile(sorted_samples, number_of_samples, percentile_rank)
+      Map.put(acc, percentile_rank, perc)
+    end)
```
This is just one pipe, so I think it can be written a bit more easily as `Enum.reduce(percentile_ranks, %{} ...`

wasnotrice replied:
CREDO!!! Fails on the line length check. Can't get that one thru CI

PragTob replied:
Huh, I see you like credo :D I might come around on not enforcing 80 char width but not quite yet :D 

Worst case I'd break the arguments to multiple lines.

devonestes replied:
It's going to be 98 characters with the formatter. That's one of the few things I disagree with, but c'est la vie! I'd be ok with changing Credo's line length check to be 98 characters now, since we know that's in the future anyway.

PragTob replied:
98? 98????? Not 100. No 98? _exahles slowly_

(for people who don't know me, I'm mostly just joking around here but am genuinely confused)

Go ahead @devonestes

wasnotrice replied:
I actually really like credo. But the hard 80-char limit is really tough in certain cases, like when you are trying to put a url in a comment

OvermindDL1 replied:
_/me always uses a 120 char limit everywhere for everything, so this 98 limit in the new formatter is blegh_  ^.^;

PragTob commented on lib/benchee/statistics/percentile.ex:
```diff
@@ -1,50 +1,67 @@
-    sorted = Enum.sort(samples)
-    rank = (percentile_number / 100) * max(0, number_of_samples + 1)
-    percentile_value(sorted, rank)
+  defp percentile(sorted_samples, number_of_samples, percentile_rank) do
```
these days we usually group the same function together without a newline inbetween, should find/create a credo rule for that :)

wasnotrice replied:
@devonestes code formatter now separates these functions, no?

PragTob replied:
what... everything I know is a lie then I believe I started adopting that style after seeing how it's done in elixir core :| I probably should play with the formatter a little...

devonestes replied:
Yep, there's always a space between functions defined with the do-newline-end syntax. If you want no line between functions, you can use `, do:` in the pure AST form.

PragTob replied:
:'(

wasnotrice replied:
@PragTob yes, some of the code formatter styles are sad, I agree.

PragTob commented on lib/benchee/statistics/percentile.ex:
```diff
@@ -1,50 +1,67 @@
+    percentile_value(sorted_samples, rank)
   end
 
   defp percentile_value(sorted, rank) when trunc(rank) == 0 do
```
With our new error throwing ways we shouldn't need this method anymore :)

PragTob replied:
Same thing for the function version below but it won't let me comment :)

wasnotrice replied:
this probably got buried in my initial comment, but I added this back as a private function mainly so we don't have to sort the list again while calculating multiple percentiles on the same list

PragTob replied:
i meant the specific `== 0` but maybe we can't get rid off it :'(

wasnotrice replied:
I'll try—it's actually a guard for the `Enum.drop` later, to make sure we have enough elements to make a match. But I'm not sure it adequately protects us. I need to write some more tests.

PragTob commented on test/benchee/statistics/percentile_test.exs:
```diff
@@ -18,7 +18,8 @@ defmodule Benchee.Statistics.PercentileTest do
 
-  @tag run: true
+  # Test data from:
+  #   http://www.itl.nist.gov/div898/handbook/prc/section2/prc262.htm
```
:heart_eyes:

wasnotrice commented:
I think this is ready to merge. I probably should have put the console formatter stuff in another commit, since we may want it to be optional, per #50. But I can take that one up next, since I've just been in that code.

PragTob commented on lib/benchee/statistics/percentile.ex:
```diff
@@ -17,38 +17,33 @@ defmodule Benchee.Statistics.Percentile do
-    when is_list(percentile_ranks) do
+    sorted_samples = Enum.sort(samples)
     percentile_ranks
+    |> List.wrap
```
Had to look this one up, nice! :)

PragTob commented on lib/benchee/statistics/percentile.ex:
```diff
@@ -64,24 +59,22 @@ defmodule Benchee.Statistics.Percentile do
+        |> Enum.reverse
+        |> hd
+        |> to_float
+    end
```
Method is a bit too long for my sensibilities :D I feel like at least the last case and maybe the first one could go into a separate function (no huge gains for the first). Wdyt?

(not totally opposed to shipping as is)

PragTob commented on test/benchee/formatters/console_test.exs:
```diff
@@ -63,12 +100,26 @@ defmodule Benchee.Formatters.ConsoleTest do
+            median: 195.5,
+            percentiles: %{99 => 300.1}
+          }
+        }
```
Nice formatting :+1:

PragTob commented on test/benchee/statistics/percentile_test.exs:
```diff
@@ -21,7 +21,74 @@ defmodule Benchee.Statistics.PercentileTest do
   end
+
+  test "an empty list raises an argument error" do
+    assert_raise ArgumentError, fn -> Percentile.percentiles([], [1]) end
```
:+1: good edge case testing!

PragTob commented on test/benchee/statistics/percentile_test.exs:
```diff
@@ -21,7 +21,74 @@ defmodule Benchee.Statistics.PercentileTest do
+      %{99 => result} = Percentile.percentiles(samples, [99])
+      assert result == 300.0
+    end
+  end
```
even better edge case testing! :)

PragTob commented:
@wasnotrice ah forgot to say it, I think we wanna display the default percentile all the time so it should be alright, only for optional more percentiles they should land in the additional ones (or so I think). Depends on how wide the output is now :)

Once we have this on master, some playing around might also be nice. I.e. 99th percentile might be a sensible default, but it might not be. Iirc Skylight uses 95th so maybe that ends up being the better default, we might wanna try and play with that a bit.

wasnotrice commented:
@PragTob good call on splitting that case up. I ended up splitting the whole thing into a set of function clauses, which makes the flow clearer, I think.

I agree that 99th percentile might not be the best default. It probably depends a lot on the data. I envision it being easy to configure a single percentile to show in the main results, and maybe a set of percentiles to display in the extended results.

The current display with `99th %` for `examples/simple_measure.exs` is 75 characters:

```
Benchmarking map...

Name           ips        average  deviation         median         99th %
map        20.54 K       48.68 μs    ±44.07%          43 μs         109 μs
```

Thanks for the great review @PragTob @devonestes 🎉

PragTob commented:
So with all this, this can go in :green_heart: :tada: 

Yeah, depending on what we want 99th percentile doesn't seem like a great default. I mean right now it seems to be more or less a worst case performance but we already have this with the `maximum` metric (we just don't display it... yet ;P). I envisioned it more as a "worst case without the outliers" sort of thing.

As for configuration, I think it'd be fair to say that the first value in the percentiles list always gets displayed. But that's for another issue.

Thanks so much @wasnotrice !
